pkg/istio-agent/xds_proxy.go

```diff
@@ -839,3 +821,21 @@ func (p *XdsProxy) initDebugInterface(port int) error {
 
 	return nil
 }
+
+// upstreamErr sends the error to upstreamError channel, and return immediately if the connection closed.
+func upstreamErr(con *ProxyConnection, err error) {
+	proxyLog.Errorf("upstream [%d] error: %v", con.conID, err)
+	select {
+	case con.upstreamError <- err:
+	case <-con.stopChan:
+	}
+}
+
+// downstreamErr sends the error to downstreamError channel, and return immediately if the connection closed.
+func downstreamErr(con *ProxyConnection, err error) {
+	proxyLog.Errorf("downstream [%d] error: %v", con.conID, err)
+	select {
+	case con.downstreamError <- err:
+	case <-con.stopChan:
```

Comment:
I think there is a subtle race condition.

1. `con.stopChan` is closed
2. `handleUpstream` (where channel is read) exits. nothing reading downstreamError anymore
3. `select` case picks case 1 instead of case 2 and blocks forever

We need to have bias to `con.stopChan` first.

Our trick in https://github.com/istio/istio/blob/1a6a789e9927397fa80ffc16624843d7b9cef63c/pilot/pkg/xds/ads.go#L329 will not work either, since we could have a race between the two `select` as well (here, we can concurrently call `downstreamErr`)

Reply:
2 cannot happen if con.stopChan is closed,because after `handleUpstream` returns then stopChan will be closed

Reply:
>We need to have bias to con.stopChan first.

Do you mean it can select `case con.downstreamError`  even con.stopChan is closed?

Reply:
Yes, go select can pick randomly https://medium.com/a-journey-with-go/go-ordering-in-select-statements-fd0ff80fd8d6

Reply:
maybe there is precedence on read vs write though.. should check

Reply:
oh wait, the downstreamErr case will not be picked since it's a unbuffered channel. So I think this is fine, right?

Reply:
I put a hold to make sure you agree, approving though - feel free to remove

Reply:
to be more clear - this would be an issue only if we made it a buffered channel of size 1

Reply:
yes, if there is no default, and none of the channels are ready, select will block until any case is ready.